File: src/registry/manifest.rs
use reqwest::Client;
use serde::Deserialize;

use super::{detect_registry, RegistryKind};

#[derive(Debug, Clone, serde::Serialize, Deserialize)]
pub struct Platform {
    pub os: String,
    pub architecture: String,
    pub variant: Option<String>,
}

// Registry token response
#[derive(Deserialize)]
struct TokenResponse {
    token: Option<String>,
    access_token: Option<String>,
}

// Manifest list response (manifest.list.v2 or OCI image index)
#[derive(Deserialize)]
struct ManifestList {
    #[serde(rename = "mediaType", default)]
    media_type: String,
    #[serde(rename = "schemaVersion")]
    schema_version: Option<u32>,
    manifests: Option<Vec<ManifestEntry>>,
}

#[derive(Deserialize)]
struct ManifestEntry {
    platform: Option<PlatformRaw>,
    digest: Option<String>,
}

#[derive(Deserialize)]
struct PlatformRaw {
    os: Option<String>,
    architecture: Option<String>,
    variant: Option<String>,
}

async fn get_registry_token(
    client: &Client,
    token_url: &str,
    service: &str,
    scope: &str,
    bearer_token: &str,
    username: &str,
    password: &str,
) -> Option<String> {
    let mut req = client
        .get(token_url)
        .query(&[("service", service), ("scope", scope)]);

    if !bearer_token.is_empty() {
        req = req.bearer_auth(bearer_token);
    } else if !username.is_empty() && !password.is_empty() {
        req = req.basic_auth(username, Some(password));
    }

    let resp = req.send().await.ok()?;
    if !resp.status().is_success() {
        return None;
    }
    let data: TokenResponse = resp.json().await.ok()?;
    data.token.or(data.access_token)
}

fn parse_manifest_platforms(data: ManifestList) -> Option<Vec<Platform>> {
    let is_list =
        data.media_type.contains("manifest.list") || data.media_type.contains("image.index");
    let is_schema_v2 = data.schema_version == Some(2);

    if !is_list && !is_schema_v2 {
        return None;
    }

    let entries = data.manifests?;
    let platforms: Vec<Platform> = entries
        .into_iter()
        .filter_map(|e| {
            let p = e.platform?;
            Some(Platform {
                os: p.os.unwrap_or_default(),
                architecture: p.architecture.unwrap_or_default(),
                variant: p.variant,
            })
        })
        .collect();

    if platforms.is_empty() {
        None
    } else {
        Some(platforms)
    }
}

async fn fetch_platforms_from_url(
    client: &Client,
    manifest_url: &str,
    auth_token: &str,
) -> Option<Vec<Platform>> {
    let accept = concat!(
        "application/vnd.docker.distribution.manifest.list.v2+json,",
        "application/vnd.oci.image.index.v1+json"
    );
    let resp = client
        .get(manifest_url)
        .header("Accept", accept)
        .bearer_auth(auth_token)
        .send()
        .await
        .ok()?;
    if !resp.status().is_success() {
        return None;
    }
    let data: ManifestList = resp.json().await.ok()?;
    parse_manifest_platforms(data)
}

async fn fetch_digest_from_url(
    client: &Client,
    manifest_url: &str,
    auth_token: &str,
) -> Option<String> {
    let accept = concat!(
        "application/vnd.docker.distribution.manifest.v2+json,",
        "application/vnd.oci.image.manifest.v1+json,",
        "application/vnd.docker.distribution.manifest.list.v2+json,",
        "application/vnd.oci.image.index.v1+json"
    );
    let resp = client
        .head(manifest_url)
        .header("Accept", accept)
        .bearer_auth(auth_token)
        .send()
        .await
        .ok()?;
    if !resp.status().is_success() {
        return None;
    }
    resp.headers()
        .get("Docker-Content-Digest")
        .and_then(|v| v.to_str().ok())
        .map(|s| s.to_string())
}

async fn fetch_platform_digest_from_url(
    client: &Client,
    manifest_url: &str,
    auth_token: &str,
    os: &str,
    arch: &str,
) -> Option<String> {
    let accept = concat!(
        "application/vnd.docker.distribution.manifest.list.v2+json,",
        "application/vnd.oci.image.index.v1+json"
    );
    let resp = client
        .get(manifest_url)
        .header("Accept", accept)
        .bearer_auth(auth_token)
        .send()
        .await
        .ok()?;
    if !resp.status().is_success() {
        return None;
    }
    let data: ManifestList = resp.json().await.ok()?;
    let entries = data.manifests?;
    for entry in entries {
        if let Some(p) = &entry.platform {
            if p.os.as_deref() == Some(os) && p.architecture.as_deref() == Some(arch) {
                return entry.digest;
            }
        }
    }
    None
}

// ---------------------------------------------------------------------------
// DockerHub helpers
// ---------------------------------------------------------------------------

fn dockerhub_image_name(image: &str) -> String {
    let name = image.strip_prefix("docker.io/").unwrap_or(image);
    let parts: Vec<&str> = name.split('/').collect();
    if parts.len() == 1 {
        format!("library/{name}")
    } else {
        name.to_string()
    }
}

async fn dockerhub_token(
    client: &Client,
    name: &str,
    username: &str,
    password: &str,
    auth_base: &str,
) -> Option<String> {
    let token_url = format!("{auth_base}/token");
    let scope = format!("repository:{name}:pull");
    get_registry_token(
        client,
        &token_url,
        "registry.docker.io",
        &scope,
        "",
        username,
        password,
    )
    .await
}

async fn ghcr_token(
    client: &Client,
    host: &str,
    path: &str,
    github_token: &str,
    registry_base: &str,
) -> Option<String> {
    if github_token.is_empty() {
        return None;
    }
    let token_url = format!("{registry_base}/token");
    let scope = format!("repository:{path}:pull");
    get_registry_token(client, &token_url, host, &scope, github_token, "", "").await
}

fn ghcr_host_and_path(image: &str) -> (String, String) {
    let image = image.trim();
    let parsed_host = if image.contains("://") {
        image
            .split("://")
            .nth(1)
            .unwrap_or("")
            .split('/')
            .next()
            .unwrap_or("")
            .split(':')
            .next()
            .unwrap_or("")
            .to_lowercase()
    } else if image.contains('/') {
        image.split('/').next().unwrap_or("").to_lowercase()
    } else {
        String::new()
    };

    let host = if parsed_host == "lscr.io" {
        "lscr.io".to_string()
    } else {
        "ghcr.io".to_string()
    };
    let path = image
        .strip_prefix(&format!("{host}/"))
        .unwrap_or(image)
        .to_string();
    (host, path)
}

// ---------------------------------------------------------------------------
// Public API
// ---------------------------------------------------------------------------

pub fn is_platform_supported(platforms: Option<&[Platform]>, os: &str, arch: &str) -> bool {
    match platforms {
        None => true,
        Some(list) => list.iter().any(|p| p.os == os && p.architecture == arch),
    }
}

pub async fn fetch_manifest_list(
    client: &Client,
    image: &str,
    tag: &str,
    dh_username: &str,
    dh_password: &str,
    github_token: &str,
) -> Option<Vec<Platform>> {
    match detect_registry(image) {
        RegistryKind::DockerHub => {
            let name = dockerhub_image_name(image);
            let token = dockerhub_token(
                client,
                &name,
                dh_username,
                dh_password,
                "https://auth.docker.io",
            )
            .await?;
            let url = format!("https://registry-1.docker.io/v2/{name}/manifests/{tag}");
            fetch_platforms_from_url(client, &url, &token).await
        }
        RegistryKind::Ghcr => {
            let (host, path) = ghcr_host_and_path(image);
            let token = ghcr_token(
                client,
                &host,
                &path,
                github_token,
                &format!("https://{host}"),
            )
            .await?;
            let url = format!("https://{host}/v2/{path}/manifests/{tag}");
            fetch_platforms_from_url(client, &url, &token).await
        }
        RegistryKind::Unknown => None,
    }
}

pub async fn fetch_digest(
    client: &Client,
    image: &str,
    tag: &str,
    dh_username: &str,
    dh_password: &str,
    github_token: &str,
) -> Option<String> {
    match detect_registry(image) {
        RegistryKind::DockerHub => {
            let name = dockerhub_image_name(image);
            let token = dockerhub_token(
                client,
                &name,
                dh_username,
                dh_password,
                "https://auth.docker.io",
            )
            .await?;
            let url = format!("https://registry-1.docker.io/v2/{name}/manifests/{tag}");
            fetch_digest_from_url(client, &url, &token).await
        }
        RegistryKind::Ghcr => {
            let (host, path) = ghcr_host_and_path(image);
            let token = ghcr_token(
                client,
                &host,
                &path,
                github_token,
                &format!("https://{host}"),
            )
            .await?;
            let url = format!("https://{host}/v2/{path}/manifests/{tag}");
            fetch_digest_from_url(client, &url, &token).await
        }
        RegistryKind::Unknown => None,
    }
}

#[allow(clippy::too_many_arguments)]
pub async fn fetch_platform_digest(
    client: &Client,
    image: &str,
    tag: &str,
    os: &str,
    arch: &str,
    dh_username: &str,
    dh_password: &str,
    github_token: &str,
) -> Option<String> {
    match detect_registry(image) {
        RegistryKind::DockerHub => {
            let name = dockerhub_image_name(image);
            let token = dockerhub_token(
                client,
                &name,
                dh_username,
                dh_password,
                "https://auth.docker.io",
            )
            .await?;
            let url = format!("https://registry-1.docker.io/v2/{name}/manifests/{tag}");
            fetch_platform_digest_from_url(client, &url, &token, os, arch).await
        }
        RegistryKind::Ghcr => {
            let (host, path) = ghcr_host_and_path(image);
            let token = ghcr_token(
                client,
                &host,
                &path,
                github_token,
                &format!("https://{host}"),
            )
            .await?;
            let url = format!("https://{host}/v2/{path}/manifests/{tag}");
            fetch_platform_digest_from_url(client, &url, &token, os, arch).await
        }
        RegistryKind::Unknown => None,
    }
}

File: app/registry/manifest.py
"""Fetch manifest lists (fat manifests) to verify multi-architecture support.

Queries the registry v2 manifest endpoint for a given image:tag and returns
the list of supported platforms. Single-arch images (no manifest list) are
treated as compatible. Results are cached per (image_name, tag) pair.
"""

import base64
from typing import Optional
from urllib.parse import urlparse

import requests

import app.http as _http
from app.config import log
from app.registry.base import detect_registry

# Module-level cache: (image_name, tag) → list[dict] | None
# list[dict]: platform entries from a multi-arch manifest list
# None: single-arch image; treat as compatible
_cache: dict[tuple[str, str], Optional[list[dict]]] = {}


def clear_cache() -> None:
    """Clear the manifest list cache. Used in tests."""
    _cache.clear()


# ---------------------------------------------------------------------------
# Internal helpers
# ---------------------------------------------------------------------------

def _get_token(realm: str, service: str, scope: str, bearer_token: str = "",
               username: str = "", password: str = "") -> Optional[str]:
    """Obtain a Bearer token from a registry token endpoint."""
    headers: dict[str, str] = {}
    if bearer_token:
        headers["Authorization"] = f"Bearer {bearer_token}"
    elif username and password:
        creds = base64.b64encode(f"{username}:{password}".encode()).decode()
        headers["Authorization"] = f"Basic {creds}"

    params = {"service": service, "scope": scope}
    try:
        resp = _http.http_session.get(realm, headers=headers, params=params, timeout=10)
        resp.raise_for_status()
        data = resp.json()
        return data.get("token") or data.get("access_token")
    except Exception as exc:
        log.warning(f"Registry token request failed ({realm}): {exc}")
        return None


def _fetch_platforms_from_url(url: str, auth_headers: dict) -> Optional[list[dict]]:
    """
    GET the manifest URL and return the platform list if it is a manifest list,
    or None if it is a single-arch manifest.
    """
    headers = {
        **auth_headers,
        "Accept": (
            "application/vnd.docker.distribution.manifest.list.v2+json,"
            "application/vnd.oci.image.index.v1+json"
        ),
    }
    try:
        resp = _http.http_session.get(url, headers=headers, timeout=15)
        resp.raise_for_status()
        data = resp.json()
        media_type = data.get("mediaType", "") or ""
        # Manifest list / OCI image index → has a manifests[] with platform info
        if "manifest.list" in media_type or "image.index" in media_type:
            return [m.get("platform", {}) for m in data.get("manifests", [])]
        # Fallback: schema v2 with a manifests[] key (e.g. some OCI indexes omit mediaType)
        if data.get("schemaVersion") == 2 and isinstance(data.get("manifests"), list):
            platforms = [m.get("platform", {}) for m in data["manifests"]]
            if platforms:
                return platforms
        # Single-arch manifest
        return None
    except requests.HTTPError as exc:
        log.warning(f"Manifest fetch HTTP error ({url}): {exc}")
        return None
    except Exception as exc:
        log.warning(f"Manifest fetch error ({url}): {exc}")
        return None


# ---------------------------------------------------------------------------
# Registry-specific fetchers
# ---------------------------------------------------------------------------

def _fetch_dockerhub_manifest_list(
    image_name: str, tag: str, username: str, password: str
) -> Optional[list[dict]]:
    name = image_name.removeprefix("docker.io/")
    parts = name.split("/")
    if len(parts) == 1:
        name = f"library/{name}"

    scope = f"repository:{name}:pull"
    token = _get_token(
        "https://auth.docker.io/token",
        service="registry.docker.io",
        scope=scope,
        username=username,
        password=password,
    )
    if not token:
        log.warning(f"DockerHub: could not obtain registry token for {name} — skipping arch check")
        return None

    url = f"https://registry-1.docker.io/v2/{name}/manifests/{tag}"
    return _fetch_platforms_from_url(url, {"Authorization": f"Bearer {token}"})


def _fetch_ghcr_manifest_list(
    image_name: str, tag: str, github_token: str
) -> Optional[list[dict]]:
    if not github_token:
        log.warning(f"GHCR: no GITHUB_TOKEN — skipping arch check for {image_name}")
        return None

    image_ref = image_name.strip()
    parsed = urlparse(image_ref)

    parsed_host = ""
    if parsed.scheme and parsed.netloc:
        parsed_host = (parsed.hostname or "").lower()
    elif "/" in image_ref:
        parsed_host = image_ref.split("/", 1)[0].lower()

    host = "lscr.io" if parsed_host == "lscr.io" else "ghcr.io"
    path = image_ref.removeprefix(f"{host}/")

    # Exchange GitHub PAT for a registry-scoped token
    scope = f"repository:{path}:pull"
    reg_token = _get_token(
        f"https://{host}/token",
        service=host,
        scope=scope,
        bearer_token=github_token,
    )
    if not reg_token:
        log.warning(f"GHCR: could not obtain registry token for {path} — skipping arch check")
        return None

    url = f"https://{host}/v2/{path}/manifests/{tag}"
    return _fetch_platforms_from_url(url, {"Authorization": f"Bearer {reg_token}"})


# ---------------------------------------------------------------------------
# Public API
# ---------------------------------------------------------------------------

def fetch_manifest_list(
    image_name: str,
    tag: str,
    dockerhub_username: str,
    dockerhub_password: str,
    github_token: str,
) -> Optional[list[dict]]:
    """Return the list of supported platforms for *image_name*:*tag*.

    Returns:
        list[dict]: platform dicts (keys: ``os``, ``architecture``, optionally ``variant``)
                    when the image has a manifest list.
        None: the image is single-arch or the manifest could not be fetched;
              the caller should treat this as compatible (current behaviour).

    Results are cached per (image_name, tag) for the lifetime of the process.
    """
    cache_key = (image_name, tag)
    if cache_key in _cache:
        return _cache[cache_key]

    registry = detect_registry(image_name)
    if registry == "dockerhub":
        result = _fetch_dockerhub_manifest_list(image_name, tag, dockerhub_username, dockerhub_password)
    elif registry == "ghcr":
        result = _fetch_ghcr_manifest_list(image_name, tag, github_token)
    else:
        log.debug(f"Unknown registry for '{image_name}' — skipping arch check")
        result = None

    _cache[cache_key] = result
    return result


def is_platform_supported(
    platforms: Optional[list[dict]],
    os: str,
    architecture: str,
) -> bool:
    """Return True if *os*/*architecture* is present in *platforms*.

    If *platforms* is ``None`` (single-arch image or manifest unavailable),
    returns ``True`` to preserve the existing (compatible) behaviour.
    """
    if platforms is None:
        return True
    return any(
        p.get("os") == os and p.get("architecture") == architecture
        for p in platforms
    )


# ---------------------------------------------------------------------------
# Digest fetching
# ---------------------------------------------------------------------------

# Cache: (image_name, tag) → digest string or None
_digest_cache: dict[tuple[str, str], Optional[str]] = {}


def clear_digest_cache() -> None:
    """Clear the digest cache. Used in tests."""
    _digest_cache.clear()


def _fetch_digest_from_url(url: str, auth_headers: dict) -> Optional[str]:
    """HEAD the manifest URL and return the Docker-Content-Digest header."""
    headers = {
        **auth_headers,
        "Accept": (
            "application/vnd.docker.distribution.manifest.v2+json,"
            "application/vnd.oci.image.manifest.v1+json,"
            "application/vnd.docker.distribution.manifest.list.v2+json,"
            "application/vnd.oci.image.index.v1+json"
        ),
    }
    try:
        resp = _http.http_session.head(url, headers=headers, timeout=15)
        resp.raise_for_status()
        return resp.headers.get("Docker-Content-Digest") or None
    except Exception as exc:
        log.warning(f"Digest fetch error ({url}): {exc}")
        return None


def _fetch_dockerhub_digest(image_name: str, tag: str, username: str, password: str) -> Optional[str]:
    name = image_name.removeprefix("docker.io/")
    parts = name.split("/")
    if len(parts) == 1:
        name = f"library/{name}"

    scope = f"repository:{name}:pull"
    token = _get_token(
        "https://auth.docker.io/token",
        service="registry.docker.io",
        scope=scope,
        username=username,
        password=password,
    )
    if not token:
        log.warning(f"DockerHub: could not obtain registry token for {name} — skipping digest fetch")
        return None

    url = f"https://registry-1.docker.io/v2/{name}/manifests/{tag}"
    return _fetch_digest_from_url(url, {"Authorization": f"Bearer {token}"})


def _fetch_ghcr_digest(image_name: str, tag: str, github_token: str) -> Optional[str]:
    if not github_token:
        log.warning(f"GHCR: no GITHUB_TOKEN — skipping digest fetch for {image_name}")
        return None

    image_ref = image_name.strip()
    parsed = urlparse(image_ref)

    parsed_host = ""
    if parsed.scheme and parsed.netloc:
        parsed_host = (parsed.hostname or "").lower()
    elif "/" in image_ref:
        parsed_host = image_ref.split("/", 1)[0].lower()

    host = "lscr.io" if parsed_host == "lscr.io" else "ghcr.io"
    path = image_ref.removeprefix(f"{host}/")

    scope = f"repository:{path}:pull"
    reg_token = _get_token(
        f"https://{host}/token",
        service=host,
        scope=scope,
        bearer_token=github_token,
    )
    if not reg_token:
        log.warning(f"GHCR: could not obtain registry token for {path} — skipping digest fetch")
        return None

    url = f"https://{host}/v2/{path}/manifests/{tag}"
    return _fetch_digest_from_url(url, {"Authorization": f"Bearer {reg_token}"})


def fetch_digest(
    image_name: str,
    tag: str,
    dockerhub_username: str,
    dockerhub_password: str,
    github_token: str,
) -> Optional[str]:
    """Return the manifest digest for *image_name*:*tag*.

    Returns:
        str: the Docker-Content-Digest value (e.g. "sha256:abc123...")
        None: could not fetch.

    Results are cached per (image_name, tag) for the lifetime of the process.
    """
    cache_key = (image_name, tag)
    if cache_key in _digest_cache:
        return _digest_cache[cache_key]

    registry = detect_registry(image_name)
    if registry == "dockerhub":
        result = _fetch_dockerhub_digest(image_name, tag, dockerhub_username, dockerhub_password)
    elif registry == "ghcr":
        result = _fetch_ghcr_digest(image_name, tag, github_token)
    else:
        log.debug(f"Unknown registry for '{image_name}' — skipping digest fetch")
        result = None

    _digest_cache[cache_key] = result
    return result


# ---------------------------------------------------------------------------
# Platform-specific digest fetching (for multi-arch manifest lists)
# ---------------------------------------------------------------------------

# Cache: (image_name, tag, os, architecture) → digest string or None
_platform_digest_cache: dict[tuple[str, str, str, str], Optional[str]] = {}


def clear_platform_digest_cache() -> None:
    """Clear the platform digest cache. Used in tests."""
    _platform_digest_cache.clear()


def _fetch_platform_digest_from_url(
    url: str, auth_headers: dict, os: str, architecture: str
) -> Optional[str]:
    """GET the manifest URL and return the digest for the specified platform."""
    headers = {
        **auth_headers,
        "Accept": (
            "application/vnd.docker.distribution.manifest.list.v2+json,"
            "application/vnd.oci.image.index.v1+json"
        ),
    }
    try:
        resp = _http.http_session.get(url, headers=headers, timeout=15)
        resp.raise_for_status()
        data = resp.json()
        media_type = data.get("mediaType", "") or ""
        manifests = None
        if "manifest.list" in media_type or "image.index" in media_type:
            manifests = data.get("manifests", [])
        elif data.get("schemaVersion") == 2 and isinstance(data.get("manifests"), list):
            manifests = data["manifests"]
        if not manifests:
            return None
        for m in manifests:
            platform = m.get("platform", {})
            if platform.get("os") == os and platform.get("architecture") == architecture:
                return m.get("digest") or None
        return None
    except requests.HTTPError as exc:
        log.warning(f"Platform digest fetch HTTP error ({url}): {exc}")
        return None
    except Exception as exc:
        log.warning(f"Platform digest fetch error ({url}): {exc}")
        return None


def _fetch_dockerhub_platform_digest(
    image_name: str, tag: str, os: str, architecture: str,
    username: str, password: str,
) -> Optional[str]:
    name = image_name.removeprefix("docker.io/")
    parts = name.split("/")
    if len(parts) == 1:
        name = f"library/{name}"

    scope = f"repository:{name}:pull"
    token = _get_token(
        "https://auth.docker.io/token",
        service="registry.docker.io",
        scope=scope,
        username=username,
        password=password,
    )
    if not token:
        log.warning(f"DockerHub: could not obtain registry token for {name} — skipping platform digest fetch")
        return None

    url = f"https://registry-1.docker.io/v2/{name}/manifests/{tag}"
    return _fetch_platform_digest_from_url(url, {"Authorization": f"Bearer {token}"}, os, architecture)


def _fetch_ghcr_platform_digest(
    image_name: str, tag: str, os: str, architecture: str, github_token: str,
) -> Optional[str]:
    if not github_token:
        log.warning(f"GHCR: no GITHUB_TOKEN — skipping platform digest fetch for {image_name}")
        return None

    image_ref = image_name.strip()
    parsed = urlparse(image_ref)

    parsed_host = ""
    if parsed.scheme and parsed.netloc:
        parsed_host = (parsed.hostname or "").lower()
    elif "/" in image_ref:
        parsed_host = image_ref.split("/", 1)[0].lower()

    host = "lscr.io" if parsed_host == "lscr.io" else "ghcr.io"
    path = image_ref.removeprefix(f"{host}/")

    scope = f"repository:{path}:pull"
    reg_token = _get_token(
        f"https://{host}/token",
        service=host,
        scope=scope,
        bearer_token=github_token,
    )
    if not reg_token:
        log.warning(f"GHCR: could not obtain registry token for {path} — skipping platform digest fetch")
        return None

    url = f"https://{host}/v2/{path}/manifests/{tag}"
    return _fetch_platform_digest_from_url(url, {"Authorization": f"Bearer {reg_token}"}, os, architecture)


def fetch_platform_digest(
    image_name: str,
    tag: str,
    os: str,
    architecture: str,
    dockerhub_username: str,
    dockerhub_password: str,
    github_token: str,
) -> Optional[str]:
    """Return the digest of the platform-specific manifest entry for *os*/*architecture*.

    Returns:
        str: the digest of the matching platform manifest (e.g. "sha256:abc123...")
        None: image is single-arch, platform not found in manifest list, or fetch failed.

    Results are cached per (image_name, tag, os, architecture).
    """
    cache_key = (image_name, tag, os, architecture)
    if cache_key in _platform_digest_cache:
        return _platform_digest_cache[cache_key]

    registry = detect_registry(image_name)
    if registry == "dockerhub":
        result = _fetch_dockerhub_platform_digest(
            image_name, tag, os, architecture, dockerhub_username, dockerhub_password
        )
    elif registry == "ghcr":
        result = _fetch_ghcr_platform_digest(image_name, tag, os, architecture, github_token)
    else:
        log.debug(f"Unknown registry for '{image_name}' — skipping platform digest fetch")
        result = None

    _platform_digest_cache[cache_key] = result
    return result
